### PR TITLE
Fix LDAP lookup

### PR DESCRIPTION
**Pull request by mitruly**
Our current LDAP authenticator and higher-level functions (eg. `authenticate` and `lookup`) return `UserDoesNotExistException` as a response when all domain authentications/look ups have been exhausted despite possible issues with the LDAP connectivity. This was fine before when we simply used LDAP authentication as a means to grant access to VinylDNS, but we need to clean up the logic if we are going to enable user locking.

Changes in this pull request:
- Limit the scenarios in which `UserDoesNotExistException` is returned to make sure that the user truly does not exist

### CODE REVIEW

File: modules/portal/app/controllers/LdapAuthenticator.scala [hunk 1]
```diff
@@ -189,34 +204,57 @@ class LdapAuthenticator(
...
+      case Nil =>
+        if (ldapSearchSuccessful)
+          Left(UserDoesNotExistException(s"[$userName] LDAP entity does not exist"))
+        else
+          Left(
+            LdapServiceException(
+              "Unable to successfully perform search in at least one LDAP domain"))
+      case h :: t =>
+        f(h).recoverWith {
+          case _: LdapServiceException => findUserDetails(t, userName, f, ldapSearchSuccessful)
```
Comment (mitruly) [inline modules/portal/app/controllers/LdapAuthenticator.scala]: `LdapServiceException` is the catch-all exception for something bad happening with the LDAP authentication.

File: modules/portal/app/controllers/LdapAuthenticator.scala [hunk 2]
```diff
@@ -104,53 +110,52 @@ object LdapAuthenticator {
       .map(searchDomain ⇒ searchDomain.organization → searchDomain.domainName)
       .toMap
 
+    def searchContext(
```
Comment (mitruly) [inline modules/portal/app/controllers/LdapAuthenticator.scala]: Abstracted from `authenticate` and `lookup` methods.

File: modules/portal/app/controllers/LdapAuthenticator.scala [hunk 3]
```diff
@@ -104,53 +110,52 @@ object LdapAuthenticator {
...
+        dirContext: DirContext,
+        organization: String,
+        lookupUserName: String,
+        lookupFailureMessage: String): Either[LdapException, LdapUserDetails] =
```
Comment (pauljamescleary) [inline modules/portal/app/controllers/LdapAuthenticator.scala]: Do we need to pass in the lookupFailure here, the two failure messages are very similar I think we can just bring them together.

File: modules/portal/app/controllers/LdapAuthenticator.scala [hunk 4]
```diff
@@ -175,7 +180,17 @@ object LdapAuthenticator {
   }
 }
 
-class UserDoesNotExistException(message: String) extends Exception(message: String)
+trait LdapException extends Exception {
```
Comment (pauljamescleary) [inline modules/portal/app/controllers/LdapAuthenticator.scala]: better to use `sealed abstract class LdapException(message: String) extends Exception(message)`

File: modules/portal/app/controllers/LdapAuthenticator.scala [hunk 5]
```diff
@@ -175,7 +180,17 @@ object LdapAuthenticator {
...
+  val message: String
+}
+
+case class UserDoesNotExistException(message: String) extends LdapException
```
Comment (pauljamescleary) [inline modules/portal/app/controllers/LdapAuthenticator.scala]: Maybe I didn't look far enough, but `case class` should always be marked `final` to ensure that pattern matches are total.

File: modules/portal/app/controllers/LdapAuthenticator.scala [hunk 6]
```diff
@@ -189,34 +204,58 @@ class LdapAuthenticator(
     serviceAccount: ServiceAccount)
     extends Authenticator {
 
+  /**
+    * Attempts to search for user in specified LDAP domains. Attempts all LDAP domains that are specified in order; in
+    * the event that user details are not found in any of the domains, returns an error based on whether at least one
```
Comment (pauljamescleary) [inline modules/portal/app/controllers/LdapAuthenticator.scala]: this comment is hard to follow "returns an error based on whether at least one LDAP search was successful"

File: modules/portal/app/controllers/LdapAuthenticator.scala [hunk 7]
```diff
@@ -189,34 +204,58 @@ class LdapAuthenticator(
...
+    */
   private def findUserDetails(
       domains: List[LdapSearchDomain],
       userName: String,
-      f: LdapSearchDomain => Try[LdapUserDetails]): Try[LdapUserDetails] = domains match {
-    case Nil => Failure(new UserDoesNotExistException(s"[$userName] LDAP entity does not exist"))
-    case h :: t => f(h).recoverWith { case _ => findUserDetails(t, userName, f) }
-  }
+      f: LdapSearchDomain => Either[LdapException, LdapUserDetails],
+      allDomainConnectionsUp: Boolean): Either[LdapException, LdapUserDetails] =
```
Comment (pauljamescleary) [inline modules/portal/app/controllers/LdapAuthenticator.scala]: what is `allDomainConnectionsUp` for?  You can do a pattern match on `case h :: Nil` on the last item in the list, and if that fails then you know that you have exhausted all domains

Comment (mitruly) [reply modules/portal/app/controllers/LdapAuthenticator.scala]: @pauljamescleary `allDomainConnectionsUp` is to only return `UserDoesNotExistException` if all LDAP `DirContext`s succeed. Updated the comment to be clearer.

Comment (pauljamescleary) [reply modules/portal/app/controllers/LdapAuthenticator.scala]: @mitruly right, I think that is what I was getting.  If you were to add a pattern match on `case h :: Nil` you would know that you were "at the end of the list"

```scala
domains match {
  case Nil =>
    // Catch just in case someone STARTS with an empty list
    Left(UserDoesNotExistException)
  case h :: Nil =>
    // We are on the last item in the list
    f(h).recoverWith(xxx)
  case h :: t =>
    // f(h). and do recursion
}
```

Comment (pauljamescleary) [reply modules/portal/app/controllers/LdapAuthenticator.scala]: So, now that everything is an either, why can't we just scan through the list, if we find the user we exit, otherwise we keep trying?

Comment (mitruly) [reply modules/portal/app/controllers/LdapAuthenticator.scala]: @pauljamescleary That makes sense - it'd be nice to get this cleaner if possible.

There were a few attributes that I was hoping to achieve with the updated `findUserDetails` behavior:
- If we initially encounter an `LdapServiceError` (ie. unexpected and non-permissible) error, we want to keep checking in the event that another domain search succeeds (existing behavior)
- If we exhaust all domain searches and didn't encounter any unexpected errors, we want to return `UserDoesNotExistException` (existing behavior)
- If we exhaust all domain searches and encountered _at least one_ unexpected error, we don't want to say the user doesn't exist; rather, we want to indicate that there was some sort of error in the lookup process (modified behavior)

Comment (pauljamescleary) [reply modules/portal/app/controllers/LdapAuthenticator.scala]: Does this work...

```scala
domains.collectFirst { case d => f(d).toOption }.getOrElse(UserDoesNotExistException(""))
```

So, we keep trying each domain, as the result of `f` will be either a Left with a failure (which we don't really care about) or a Right with the actual user.

When we hit the first user, we would stop looking and return success

If we get to the end of the list, we would wind up with `None`, so we just say `UserDoesNotExist`

Comment (pauljamescleary) [reply modules/portal/app/controllers/LdapAuthenticator.scala]: @mitruly I see, you want to not necessarily fall back to `UserDoesNotExistException` if there was a failure.

So, you want to stop as soon as you find the user OR you get an ldap service exception (no need to keep going as your one error may be the one that has the user).

```
domains.collectFirst { case d =>
  f(d) match {
    case LdapServiceException => Some(Left(LdapServiceException))
    case user => Some(Right(user))
    case _ => None
  }
}
```

A little sloppy, but you can still use `collectFirst`.  As soon as it yields a `Some` you stop

Comment (pauljamescleary) [reply modules/portal/app/controllers/LdapAuthenticator.scala]: @mitruly I am good with your change after tinkering with it and understanding better what you are doing.

Going to resolve this conversation 👍

File: modules/portal/app/controllers/LdapAuthenticator.scala [hunk 8]
```diff
@@ -189,34 +191,59 @@ class LdapAuthenticator(
...
 
-  def authenticate(username: String, password: String): Try[LdapUserDetails] =
-    findUserDetails(searchBase, username, authenticator.authenticate(_, username, password))
+  def authenticate(username: String, password: String): Either[LdapException, LdapUserDetails] =
+    findUserDetails(searchBase, username, authenticator.authenticate(_, username, password), true)
 
-  def lookup(username: String): Try[LdapUserDetails] =
-    findUserDetails(searchBase, username, authenticator.lookup(_, username, serviceAccount))
+  def lookup(username: String): Either[LdapException, LdapUserDetails] =
+    findUserDetails(searchBase, username, authenticator.lookup(_, username, serviceAccount), true)
```
Comment (rebstar6) [inline modules/portal/app/controllers/LdapAuthenticator.scala]: is it possible to start up vinyl with an empty list for LDAP auths?

if so, this feels risky when you get to locking (and maybe handled in next pr, if so thats fine). My worry is that if you use oidc and this gets triggered with no ldap config info, all users will get locked.

this said, `true` is accurate here, i just wonder if we should have one more error that is strictly "NoLdapDomainsConfigured" for the non-configged case

File: modules/portal/app/controllers/LdapAuthenticator.scala [hunk 9]
```diff
@@ -189,34 +191,59 @@ class LdapAuthenticator(
...
+        if (allDomainConnectionsUp)
+          Left(UserDoesNotExistException(s"[$userName] LDAP entity does not exist"))
+        else
+          Left(
+            LdapServiceException(
+              "Unable to successfully perform search in at least one LDAP domain"))
+      case h :: t =>
+        f(h).recoverWith {
+          case _: LdapServiceException => findUserDetails(t, userName, f, false)
+          case _ => findUserDetails(t, userName, f, allDomainConnectionsUp)
```
Comment (rebstar6) [inline modules/portal/app/controllers/LdapAuthenticator.scala]: should the defaulting here not go the other way? seems to me like `allDomainConnectionsUp` should only stay true in the case that you get `UserDoesNotExistException` and fall to false in all others.

thinking about errors from config again, if your service account doesnt have access, createContext fails with InvalidCredentials...which has nothing to do with the user being looked up and shouldnt trigger locking

Comment (mitruly) [reply modules/portal/app/controllers/LdapAuthenticator.scala]: @rebstar6 That's a good point; updated to only have `UserDoesNotExistException` maintain that all domain connections are up.

File: modules/portal/app/controllers/LdapAuthenticator.scala (hunk 8 above)
Comment (mitruly) [reply modules/portal/app/controllers/LdapAuthenticator.scala]: I can definitely handle this in https://github.com/vinyldns/vinyldns/pull/524
